src/connector/src/parser/scalar_adapter.rs

```diff
@@ -184,6 +184,65 @@ impl ScalarAdapter<'_> {
             }
         }
     }
+
+    // return scalar and a flag indicating if the list should fallback to NULL
+    pub fn into_scalar_in_list(self, ty: DataType) -> (Option<ScalarImpl>, bool) {
```

Comment:
This `(Option<ScalarImpl>, bool)` interface is not rust idiom. It feels like C or Go and corresponds to `Option<Option<ScalarImpl>>` in Rust because `true` always implies `None` (i.e. `(Some(_), true)` is never returned).

It is true `Option<Option<ScalarImpl>>` would be less readable and I am not recommending it either. Refer to a1e70e64af517d93ca878ac74f33288601cab043 (`pgcdc-list-numeric-a`) and see how to handle the early-return.

Reply:
updated PTAL🙏